### ISSUE TITLE
UUID endian issue

### ISSUE THREAD

**Issue by gielfeldt** [nicdex/node-eventstore-client#46]
The eventid seems to get endian-swapped (or rather, it doesn't) on the way in & out of the eventstore.

`
%> node test.js
`

```
Appending...
Connected to eventstore at 172.18.0.5:1113
Stored event: 7dffd71b-c177-49ae-89ae-46a5998533d0
Look for it at: http://localhost:2113/web/index.html#/streams/testStream
```

`
%> curl -i -H "Accept: application/json" 'http://localhost:2113/streams/testStream/head/backward/1?embed=body'
`

```
{
  "title": "Event stream 'testStream'",
  "id": "http://localhost:2113/streams/testStream",
  "updated": "2017-08-31T13:21:11.807372Z",
  "streamId": "testStream",
  "author": {
    "name": "EventStore"
  },
  "headOfStream": true,
  "selfUrl": "http://localhost:2113/streams/testStream",
  "eTag": "9;-2060438500",
  "links": [
    {
      "uri": "http://localhost:2113/streams/testStream",
      "relation": "self"
    },
    {
      "uri": "http://localhost:2113/streams/testStream/head/backward/1",
      "relation": "first"
    },
    {
      "uri": "http://localhost:2113/streams/testStream/0/forward/1",
      "relation": "last"
    },
    {
      "uri": "http://localhost:2113/streams/testStream/8/backward/1",
      "relation": "next"
    },
    {
      "uri": "http://localhost:2113/streams/testStream/10/forward/1",
      "relation": "previous"
    },
    {
      "uri": "http://localhost:2113/streams/testStream/metadata",
      "relation": "metadata"
    }
  ],
  "entries": [
    {
      "eventId": "1bd7ff7d-77c1-ae49-89ae-46a5998533d0",
      "eventType": "testEvent",
      "eventNumber": 9,
      "data": "{\"a\":0.04060175919930331,\"b\":\"7bfe63f0-4529-4c61-8cc6-594fcca47720\"}",
      "streamId": "testStream",
      "isJson": true,
      "isMetaData": false,
      "isLinkMetaData": false,
      "positionEventNumber": 9,
      "positionStreamId": "testStream",
      "title": "9@testStream",
      "id": "http://localhost:2113/streams/testStream/9",
      "updated": "2017-08-31T13:21:11.807372Z",
      "author": {
        "name": "EventStore"
      },
      "summary": "testEvent",
      "links": [
        {
          "uri": "http://localhost:2113/streams/testStream/9",
          "relation": "edit"
        },
        {
          "uri": "http://localhost:2113/streams/testStream/9",
          "relation": "alternate"
        }
      ]
    }
  ]
}
```

eventid allegedly sent by node-evenstore-client:
`7dffd71b-c177-49ae-89ae-46a5998533d0`
eventid when using the http interface:
`1bd7ff7d-77c1-ae49-89ae-46a5998533d0`

The uuid-parse module, which node-eventstore-client depends on, also seem to have had this issue raised: https://github.com/zefferus/uuid-parse/issues/2

However, it looks like development of that module has stopped, or at least has very little traction. I'm not sure if there are other uuid modules, that supports different endian modes, but alternatively here's a little hacked version of uuid-parse.js:

```
'use strict';

// Maps for number <-> hex string conversion
var _byteToHex = [];
var _hexToByte = {};
for (var i = 0; i < 256; i++) {
  _byteToHex[i] = (i + 0x100).toString(16).substr(1);
  _hexToByte[_byteToHex[i]] = i;
}

// **`parse()` - Parse a UUID into it's component bytes**
function parse(s, buf, offset) {
  const i = (buf && offset) || 0;
  offset = i;
  let ii = 0;

  buf = buf || Buffer.alloc(16 + i);
  s.toLowerCase().replace(/[0-9a-f]{2}/g, function(oct) {
    if (ii < 16) { // Don't overflow!
      buf[i + ii++] = _hexToByte[oct];
    }
  });

  // Zero out remaining bytes if string was short
  while (ii < 16) {
    buf[i + ii++] = 0;
  }

  // Endian-swap hack...
  var buf2 = Buffer.from(buf);
  buf[offset+0] = buf2[offset+3];
  buf[offset+1] = buf2[offset+2];
  buf[offset+2] = buf2[offset+1];
  buf[offset+3] = buf2[offset+0];
  buf[offset+4] = buf2[offset+5];
  buf[offset+5] = buf2[offset+4];
  buf[offset+6] = buf2[offset+7];
  buf[offset+7] = buf2[offset+6];
  
  return buf;
}

// **`unparse()` - Convert UUID byte array (ala parse()) into a string**
function unparse(buf, offset) {
  let i = offset || 0;

  // Endian-swap hack...
  var buf2 = Buffer.from(buf);
  buf[i+0] = buf2[i+3];
  buf[i+1] = buf2[i+2];
  buf[i+2] = buf2[i+1];
  buf[i+3] = buf2[i+0];
  buf[i+4] = buf2[i+5];
  buf[i+5] = buf2[i+4];
  buf[i+6] = buf2[i+7];
  buf[i+7] = buf2[i+6];

  const bth = _byteToHex;
  return  bth[buf[i++]] + bth[buf[i++]] +
          bth[buf[i++]] + bth[buf[i++]] + '-' +
          bth[buf[i++]] + bth[buf[i++]] + '-' +
          bth[buf[i++]] + bth[buf[i++]] + '-' +
          bth[buf[i++]] + bth[buf[i++]] + '-' +
          bth[buf[i++]] + bth[buf[i++]] +
          bth[buf[i++]] + bth[buf[i++]] +
          bth[buf[i++]] + bth[buf[i++]];
}

module.exports = {
  parse,
  unparse
};
```

**Comment by gielfeldt** [comment]
For ease of use, I made this: https://www.npmjs.com/package/uuid-parse-le

**Comment by nicdex** [comment]
The issue is not endianess. otherwise the whole uuid would be reversed. It has to do with how Microsoft (because eventstore is in .net) stores it's uuid aka guid. So it's a guid vs uuid bytes storage issue. I will use guid instead of uuid for parsing/unparsing eventId.
See https://en.wikipedia.org/wiki/Universally_unique_identifier#Encoding (the difference in encoding is partial little-endian, so you were partly right)

**Comment by gielfeldt** [comment]
Well yes, it is an endian issue, but only the first three parts differs, since the last part is a byte array where endianess is meaningless. I did of course visit that particular wiki-page during my investigation. It didn't occur to me that eventstore uses guid instead of uuid, before you mentioned it, and that the difference is basically the subtle endianess.

I came across this page, claiming that some bits hold a special meaning. Not sure if this applies in practise at all though wrt eventstore: https://www.quora.com/What-is-the-difference-between-UUID-and-GUID

Thanks for addressing this. GUID-packages on npmjs seems a bit scarce though. Perhaps I should just rename the uuid-parse-le to guid-parse? Is there anything else from a "guid-parsing" besides the endianess that we need?
